hw/xbox/mcpx_apu.c

```diff
@@ -1394,13 +1422,13 @@ static void process_voice(MCPXAPUState *d,
             for(unsigned int channel = 0; channel < channels; channel++) {
                 switch(sample_size) {
                 case NV_PAVS_VOICE_CFG_FMT_SAMPLE_SIZE_U8:
-                    samples[channel][i] = ldub_phys(&address_space_memory, addr);
+                    samples[channel][i] = ldub_phys(&address_space_memory, addr) << 24;
                     break;
                 case NV_PAVS_VOICE_CFG_FMT_SAMPLE_SIZE_S16:
-                    samples[channel][i] = (int16_t)lduw_le_phys(&address_space_memory, addr);
+                    samples[channel][i] = (int16_t)lduw_le_phys(&address_space_memory, addr) << 16;
                     break;
                 case NV_PAVS_VOICE_CFG_FMT_SAMPLE_SIZE_S24:
-                    samples[channel][i] = (int32_t)(ldl_le_phys(&address_space_memory, addr) << 8) >> 8;
+                    samples[channel][i] = (int32_t)(ldl_le_phys(&address_space_memory, addr) ) << 8;
```

Comment:
I assume I've used 24 bit here, because that's what hardware tests suggested. The DSP also receives 24 bit samples - I'm not sure how it expects them to be packaged.. but I assume in the lower 24 bit.

Reply:
Agree, we should stick to 24b here. I scaled them up here for 32-bit system audio output, but we should do that instead when we mix down for the VP audio sink.